File: services/entity-resolution-service/src/fusion_base/domain/engine/ml_matcher.rs
use crate::models::cluster::{EntityRecord, MatchEvidence};

use super::comparator::{compare_values, normalize_phone};

pub fn score_candidate(left: &EntityRecord, right: &EntityRecord, evidence: &MatchEvidence) -> f32 {
    let name_score = compare_values("fuzzy", &left.display_name, &right.display_name);
    let email_score = compare_values(
        "email_exact",
        &attribute(left, "email"),
        &attribute(right, "email"),
    );
    let phone_score = if normalize_phone(&attribute(left, "phone"))
        == normalize_phone(&attribute(right, "phone"))
    {
        1.0
    } else {
        0.0
    };
    let company_score = compare_values(
        "fuzzy",
        &attribute(left, "company"),
        &attribute(right, "company"),
    );
    let city_score = compare_values("exact", &attribute(left, "city"), &attribute(right, "city"));

    (0.45 * evidence.rule_score
        + 0.2 * name_score
        + 0.15 * email_score
        + 0.1 * phone_score
        + 0.05 * company_score
        + 0.05 * city_score)
        .clamp(0.0, 1.0)
}

pub fn blend_scores(rule_score: f32, ml_score: f32) -> f32 {
    (0.65 * rule_score + 0.35 * ml_score).clamp(0.0, 1.0)
}

fn attribute(record: &EntityRecord, field: &str) -> String {
    record
        .attributes
        .get(field)
        .and_then(|value| value.as_str())
        .unwrap_or_default()
        .to_string()
}

File: services/entity-resolution-service/internal/domain/engine/ml_matcher.go
package engine

import "github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"

// ScoreCandidate ports `ml_matcher::score_candidate`.
func ScoreCandidate(left, right models.EntityRecord, evidence models.MatchEvidence) float32 {
	nameScore := CompareValues("fuzzy", left.DisplayName, right.DisplayName)
	emailScore := CompareValues("email_exact", attribute(left, "email"), attribute(right, "email"))
	phoneScore := float32(0)
	if NormalizePhone(attribute(left, "phone")) == NormalizePhone(attribute(right, "phone")) {
		phoneScore = 1.0
	}
	companyScore := CompareValues("fuzzy", attribute(left, "company"), attribute(right, "company"))
	cityScore := CompareValues("exact", attribute(left, "city"), attribute(right, "city"))

	return clamp01(0.45*evidence.RuleScore +
		0.20*nameScore +
		0.15*emailScore +
		0.10*phoneScore +
		0.05*companyScore +
		0.05*cityScore)
}

// BlendScores ports `ml_matcher::blend_scores`.
func BlendScores(ruleScore, mlScore float32) float32 {
	return clamp01(0.65*ruleScore + 0.35*mlScore)
}

func attribute(record models.EntityRecord, field string) string {
	v, ok := record.Attributes[field]
	if !ok {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return ""
	}
	return s
}
